### PR TITLE
borrow device memory for AUTO:GPU case to avoid performance gap

### PR DESCRIPTION
**Pull request by songbell**
Signed-off-by: fishbell <bell.song@intel.com>

### Details:
 - borrow device friendly blob if applicable, avoid performance gap in AUTO:GPU case
 - *...*

### Tickets:
 - https://jira.devtools.intel.com/browse/CVS-69728

### CODE REVIEW

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 1]
```diff
@@ -380,6 +380,16 @@ InferenceEngine::IInferRequestInternal::Ptr MultiDeviceExecutableNetwork::Create
     InferenceEngine::SoIInferRequestInternal request_to_share_blobs_with;
 
     if (_workModeIsAUTO) {
+        if (!_networkFirstReady && _networkActualNeeded) {
```
Comment (myshevts) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:383]: this (and the "if" before) are the only diff vs original loop below.

**Review by myshevts — COMMENTED** [review]
Overall, this is not the change we need I guess- with the suggested logic the blobs are borrowed only for the requests created AFTER the actual network is loaded....so to be efficient the app should wait for that.

the original idea was to create the device-friendly blobs  (simialr to https://github.com/openvinotoolkit/openvino/blob/e57a455b9254454b1821af08327d758063ee8a1a/inference-engine/samples/benchmark_app/remote_blobs_filling.cpp#L101).

What's need to be adressed: 

- the (defualt) context (unlees the LoadNetwork for AUTO was called with context- whihc is not implemented at the moment) should be used
- the entire OpenCL depenendcies (and general "GPU"  specifics) should be avoided

So I wish we have the **make_shared_blob that accepts the just the tensor's desc (and maybe the context) and returns the device-friendly blob**, hiding the device-specific specifics like OCL inside....the gpu::make_shared_blob is too low-level and exposes too much to be used transparently with other devices. @vladimir-paramuzov what do you think?

**Comment by vladimir-paramuzov** [comment]
@myshevts GPU device blobs are not mappable on CPU anymore, but we'll add a feature for GPU-friendly host blob allocation (with usm_host ptr) within 42664 task which would be more efficient than regular host memory

**Comment by myshevts** [comment]
thanks, lets's wait for the CVS-42664 first... @songbell - again, the PR doesn't solve the copies issue if CPU loads the network first and starts the inference (so that all requests/blobs are created as CPU-friendly)

File: inference-engine/src/multi_device/multi_device_exec_network.cpp (hunk 1 above)
Comment (songbell) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:383]: Hi Maxim, thanks for the clear explanation.
1) shall we need to handle the context for different hw like GPU/VPU, they have separate context related, openCL implementation for GPU which may involve opencl dependencies as you mentioned, and for vpu, the VPU context. 
2) with make_shared_blob(desc, context) in the future, does the context need to be shared between loadNetwork to the hw device and make_shared_blob?

Comment (myshevts) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:383]: >>with make_shared_blob(desc, context) in the future, does the context need to be shared between loadNetwork to the hw device and make_shared_blob?

Yes, and tis is why the LoadNetwork flavor that accepts conetxt will be likely needed

**Comment by myshevts** [comment]
Yet, I have just realized that you are trying to avoid the copies only for the AUTO:GPU case (so that AUTO always loads the GPU first)... this makes sense...

**Comment by songbell** [comment]
> Yet, I have just realized that you are trying to avoid the copies only for the AUTO:GPU case (so that AUTO always loads the GPU first)... this makes sense...

yes, the solution for the hotswap(cpu+gpu case) is been planned for local CI first and will update you when ready.

**Comment by wangleis** [comment]
@vladimir-paramuzov Just for clarification, there are two PR for GPU shared blob in two different cases:

1. This PR is created for AUTO:GPU case which GPU is initial inference device. No hot-swapping will happen.
2. For AUTO:CPU,GPU which CPU is initial inference device and hot-swapping happen after GPU ready, Bell will create another PR for CVS-42664. 

Could we merge this PR to solve performance validation on comparison between GPU and AUTO:GPU?

**Comment by myshevts** [comment]
> Could we merge this PR to solve performance validation on comparison between GPU and AUTO:GPU?
Yes

**Review by myshevts — APPROVED** [review]

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 2]
```diff
@@ -380,6 +380,11 @@ InferenceEngine::IInferRequestInternal::Ptr MultiDeviceExecutableNetwork::Create
...
+            auto& dev_requests = _workerRequests[_acceleratorDevice.deviceName];
+            auto index = num % dev_requests.size();
```
[the remark this answers is not in the document]
Comment (songbell) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:385]: yes, this is a potential bug which may mess up the inputs of different auto infer request, I will submit new PR for this, for extra user auto infer request, just stick to cpu memory as it is.

**Pull request merged**